### PR TITLE
FrontEnd OpExtension

### PR DESCRIPTION
**Pull request by itikhono**
### Details:
 - *Implementation of OpExtension class*
     OpExtension is a helper class to create conversion rules from framework operation to openvino operation.
` c++ syntax:`
     1) use ov Core class ov::OpExtension 

       class CustomOp : public ov::op::Op {
       public:
            ... // <-- some declaration and definition to define a new ov operation
            // this macro registers a new Converter in ONNXFrontEnd for "fw_op_name" operation, "ov_attribute1"  
            // will be  mapped to "fw_attribute1", "ov_attribute_2" will be set to 1
            OPENVINO_FRAMEWORK_MAP(onnx, "fw_op_name",  {{"ov_attribute1", "fw_attribute1"}}, {{"ov_attribute_2", 1}}) 
           // the same for tensorflow FrontEnd
            OPENVINO_FRAMEWORK_MAP(onnx, "fw_op_name",  {{"ov_attribute1", "fw_attribute1"}}, {{"ov_attribute_2", 1}}) 
       };
       Core ie;
       ie.add_extension(ov::OpExtension(CustomOp));

    2) use ov::frontend::OpExtension 
```
       Core ie;
       // this extension maps "fw_op_name" framework operation to existed OpenVINO "ov_op_name" operation
      // "ov_attribute1"  will be  mapped to "fw_attribute1", "ov_attribute_2" will be set to 1
       ie.add_extension(ov::frontend::OpExtension<>("ov_op_name", "fw_op_name", {{"ov_attribute1", "fw_attribute1"}}, {{"ov_attribute_2", 1}}));
```

  `  python syntax:`

```
    from openvino.frontend import OpExtension
    from openvino.runtime import Core

    ie = Core()
    # the same as in c++, this extension maps "fw_op_name" framework operation to existed OpenVINO "ov_op_name" 
    #operation
    # "ov_attribute1"  will be  mapped to "fw_attribute1", "ov_attribute_2" will be set to 1
    ie.add_extension(OpExtension("ov_op_name", "fw_op_name", {"ov_attribute_1": "fw_attribute_1"},
                                 {"ov_attribute_str": 1})
```

 - *Python bindings*
 - *Unit tests*

### Tickets:
 - *68199*
- *69478*
- *68219*

### CODE REVIEW

File: src/bindings/python/src/openvino/frontend/tensorflow/__init__.py [hunk 1]
```diff
@@ -14,6 +14,7 @@
 
 try:
     from openvino.frontend.tensorflow.py_tensorflow_frontend import ConversionExtensionTensorflow as ConversionExtension
+    from openvino.frontend.tensorflow.py_tensorflow_frontend import OpExtensionTensorflow as OpExtension
```
Comment (ilyachur) [inline src/bindings/python/src/openvino/frontend/tensorflow/__init__.py:17]: Is it possible that we can have ConversionExtension but don't have OpExtension?

File: src/bindings/python/src/pyopenvino/frontend/node_context.cpp [hunk 2]
```diff
@@ -58,7 +58,6 @@ void regclass_frontend_NodeContext(py::module m) {
         [=](NodeContext& self, const std::string& name, const py::object& default_value, const py::object& dtype)
             -> py::object {
             auto any = self.get_attribute_as_any(name);
-            auto module = py::module_::import("openvino.runtime");
```
Comment (ilyachur) [inline src/bindings/python/src/pyopenvino/frontend/node_context.cpp:61]: Why did you remove it?

File: src/frontends/common/include/openvino/frontend/extension/op.hpp [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+                                     return res;
+                                 };
+
+                                 // Expected formats:
+                                 // opsetN::OpName
+                                 // opsetN.OpName
+                                 // OpName
+                                 std::string opset_name;
+                                 std::string op_name;
+                                 auto cnt_colons = std::count(ov_type_name.begin(), ov_type_name.end(), ':');
```
Comment (itikhono) [inline src/frontends/common/include/openvino/frontend/extension/op.hpp:140]: maybe regexp is better

File: src/core/include/openvino/core/attribute_adapter.hpp [hunk 4]
```diff
@@ -34,6 +35,12 @@ class OPENVINO_API ValueAccessor<void> {
     /// as_type.
     virtual const DiscreteTypeInfo& get_type_info() const = 0;
     virtual ~ValueAccessor() = default;
+    virtual ov::Any get_as_any() {
```
Comment (ilyachur) [inline src/core/include/openvino/core/attribute_adapter.hpp:38]: Why do we need this?
Maybe should we throw ov::Exception?

File: src/core/include/openvino/core/op_extension.hpp [hunk 5]
```diff
@@ -33,12 +33,39 @@ class OPENVINO_API BaseOpExtension : public Extension {
...
 
+    /**
+     * @brief Returns extensions that should be registered together with this extension class object.
+     *
+     * Attached extensions may include frontend extensions that OpenVINO op to framework ops or necessary
+     * transformations that should be applied to the network which consist of target op.
+     *
+     * @return
+     */
+    virtual std::vector<ov::Extension::Ptr> get_attached_extensions() const {
```
Comment (ilyachur) [inline src/core/include/openvino/core/op_extension.hpp:44]: Why do we need to extend base class?

File: src/core/include/openvino/core/op_extension.hpp [hunk 6]
```diff
@@ -33,12 +33,39 @@ class OPENVINO_API BaseOpExtension : public Extension {
...
+
     /**
      * @brief Destructor
      */
     ~BaseOpExtension() override;
 };
 
+namespace detail {
+
+#define OV_COLLECT_ATTACHED_EXTENSIONS(FRAMEWORK)                                                         \
```
Comment (ilyachur) [inline src/core/include/openvino/core/op_extension.hpp:56]: Why do we have framework specific in this header?

File: src/core/include/openvino/opsets/opset.hpp [hunk 7]
```diff
@@ -136,4 +136,5 @@ const OPENVINO_API OpSet& get_opset5();
 const OPENVINO_API OpSet& get_opset6();
 const OPENVINO_API OpSet& get_opset7();
 const OPENVINO_API OpSet& get_opset8();
+const OPENVINO_API OpSet& get_opset_by_name(const std::string& opset_name);
```
Comment (ilyachur) [inline src/core/include/openvino/opsets/opset.hpp:139]: Do we need this helper in public API?

File: src/bindings/python/src/pyopenvino/frontend/node_context.cpp (hunk 2 above)
Comment (itikhono) [reply src/bindings/python/src/pyopenvino/frontend/node_context.cpp:61]: unused variable, it was added in my ConversionExtension PR for some experiments

File: src/bindings/python/src/openvino/frontend/tensorflow/__init__.py (hunk 1 above)
Comment (itikhono) [reply src/bindings/python/src/openvino/frontend/tensorflow/__init__.py:17]: OpExtension is derived from ConversionExtension, so they both should exist

File: src/frontends/common/include/openvino/frontend/extension/op.hpp (hunk 3 above)
Comment (itikhono) [reply src/frontends/common/include/openvino/frontend/extension/op.hpp:140]: regexp looks more confusing and doesn't reduce the amount of code

File: src/core/include/openvino/opsets/opset.hpp (hunk 7 above)
Comment (itikhono) [reply src/core/include/openvino/opsets/opset.hpp:139]: discussed offline
this helper will be moved for python bindings only

File: src/core/include/openvino/core/attribute_adapter.hpp (hunk 4 above)
Comment (itikhono) [reply src/core/include/openvino/core/attribute_adapter.hpp:38]: this method is needed to handle custom user attributes as ov::Any
user can map attribute names and values and do not specify a concrete type, e.g.
`{ "attr_name_1" : 5, "attr_name_2" : "string", "attr_name_3" : enum}`
actually it's map<std::string, **ov::Any**>

I will add more tests to cover most common types

File: src/core/include/openvino/core/op_extension.hpp (hunk 5 above)
Comment (itikhono) [reply src/core/include/openvino/core/op_extension.hpp:44]: discussed offline, OpExtensionBase can be extended

File: src/core/include/openvino/core/op_extension.hpp (hunk 6 above)
Comment (itikhono) [reply src/core/include/openvino/core/op_extension.hpp:56]: discussed offline
I will try to move fw mention to frontend part

File: src/core/include/openvino/opsets/opset.hpp (hunk 7 above)
Comment (itikhono) [reply src/core/include/openvino/opsets/opset.hpp]: the helper has been moved to frontend/extensions/op.hpp

File: src/core/include/openvino/core/attribute_adapter.hpp (hunk 4 above)
Comment (itikhono) [reply src/core/include/openvino/core/attribute_adapter.hpp:38]: the tests have been added
changed to ov::Exception

File: src/core/include/openvino/core/op_extension.hpp (hunk 6 above)
Comment (itikhono) [reply src/core/include/openvino/core/op_extension.hpp:53]: As a result of discussion we decided investigate this as part of separate ticket.
ticket: 77980

File: src/core/include/openvino/core/op_extension.hpp [hunk 8]
```diff
@@ -33,12 +33,36 @@ class OPENVINO_API BaseOpExtension : public Extension {
...
 
+    /**
+     * @brief Returns extensions that should be registered together with this extension class object.
+     *
+     * Attached extensions may include frontend extensions that OpenVINO op to framework ops or necessary
+     * transformations that should be applied to the network which consist of target op.
+     *
+     * @return
+     */
+    virtual std::vector<ov::Extension::Ptr> get_attached_extensions() const = 0;
```
Comment (ilya-lavrenov) [inline src/core/include/openvino/core/op_extension.hpp:44]: should not `attached_extensions` be registered before current extension is added to the FE / Core? Instead of hierarchy of extensions

Comment (itikhono) [reply src/core/include/openvino/core/op_extension.hpp:44]: What do you mean? 
To support this syntax, when we register a new frontend::OpExtension via OPENVINO_FRAMEWORK_MAP(...) macro we need to collect all in core::OpExtension method, because the user type, in this case CustomOp, is known only on this level. 
image

full pipeline looks like that:
UserClass - wrapped to -> core::OpExtension(UserClass) - casted to -> core::BaseOpExtension  - calls -> virtual method attached_extension() - collects -> frontend::OpExtension registered via OPENVINO_FRAMEWORK_MAP 
after that frontend::OpExtension are passed to each FrontEnd

File: src/bindings/python/src/pyopenvino/frontend/extension.cpp [hunk 9]
```diff
@@ -118,3 +119,37 @@ void regclass_frontend_ProgressReporterExtension(py::module m) {
...
+    py::class_<OpExtension<void>, std::shared_ptr<OpExtension<void>>, ConversionExtension> ext(m,
+                                                                                               "OpExtension",
+                                                                                               py::dynamic_attr());
+
+    ext.def(py::init([](const std::string& fw_type_name,
+                        const std::map<std::string, std::string>& attr_names_map,
+                        const std::map<std::string, py::object>& attr_values_map) {
+                std::map<std::string, ov::Any> any_map;
+                for (const auto& it : attr_values_map) {
+                    any_map[it.first] = it.second;
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/bindings/python/src/pyopenvino/frontend/extension.cpp:133]: I will double check it today

File: src/bindings/python/src/pyopenvino/frontend/tensorflow/extension.cpp [hunk 10]
```diff
@@ -39,3 +40,37 @@ void regclass_frontend_tensorflow_ConversionExtension(py::module m) {
...
+                for (const auto& it : attr_values_map) {
+                    any_map[it.first] = it.second;
+                }
+                return std::make_shared<OpExtension<void>>(ov_type_name, fw_type_name, attr_names_map, any_map);
+            }),
+            py::arg("ov_type_name"),
+            py::arg("fw_type_name"),
+            py::arg("attr_names_map") = std::map<std::string, std::string>(),
+            py::arg("attr_values_map") = std::map<std::string, py::object>());
+}
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/bindings/python/src/pyopenvino/frontend/tensorflow/extension.cpp]: fixed

File: src/bindings/python/src/pyopenvino/frontend/onnx/extension.cpp [hunk 11]
```diff
@@ -41,3 +42,37 @@ void regclass_frontend_onnx_ConversionExtension(py::module m) {
...
+        for (const auto& it : attr_values_map) {
+            any_map[it.first] = it.second;
+        }
+        return std::make_shared<OpExtension<void>>(ov_type_name, fw_type_name, attr_names_map, any_map);
+    }),
+            py::arg("ov_type_name"),
+            py::arg("fw_type_name"),
+            py::arg("attr_names_map") = std::map<std::string, std::string>(),
+            py::arg("attr_values_map") = std::map<std::string, py::object>());
+}
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/bindings/python/src/pyopenvino/frontend/onnx/extension.cpp]: fixed

File: src/bindings/python/src/pyopenvino/frontend/extension.cpp [hunk 12]
```diff
@@ -118,3 +119,37 @@ void regclass_frontend_ProgressReporterExtension(py::module m) {
...
+                }
+                return std::make_shared<OpExtension<void>>(ov_type_name, fw_type_name, attr_names_map, any_map);
+            }),
+            py::arg("ov_type_name"),
+            py::arg("fw_type_name"),
+            py::arg("attr_names_map") = std::map<std::string, std::string>(),
+            py::arg("attr_values_map") = std::map<std::string, py::object>());
+}
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/bindings/python/src/pyopenvino/frontend/extension.cpp]: fixed

File: src/frontends/tensorflow/src/frontend.cpp [hunk 13]
```diff
@@ -273,7 +273,17 @@ bool FrontEnd::supported_impl(const std::vector<ov::Any>& variants) const {
         if (ov::util::ends_with(model_path, suffix.c_str())) {
             return true;
         }
-    } else if (variants[0].is<GraphIterator::Ptr>()) {
+    }
+#if defined(OPENVINO_ENABLE_UNICODE_PATH_SUPPORT) && defined(_WIN32)
+    else if (variants[0].is<std::wstring>()) {
```
Comment (itikhono) [inline src/frontends/tensorflow/src/frontend.cpp:278]: @rkazants @ilyachur could you look at these changes?

File: src/bindings/python/src/pyopenvino/frontend/tensorflow/extension.cpp [hunk 14]
```diff
@@ -11,6 +11,7 @@
 
 #include "openvino/frontend/extension/conversion.hpp"
```
Comment (ilyachur) [inline src/bindings/python/src/pyopenvino/frontend/tensorflow/extension.cpp:12]: Could you update copyright in all modified files?

File: src/common/util/include/openvino/util/common_util.hpp [hunk 15]
```diff
@@ -93,6 +93,20 @@ inline bool ends_with(const std::string& src, const char* with) {
...
+ * @param src - string/wstring to check
+ * @param with - given substring
+ * @return true if string end with given substring
+ */
+template <typename T>
+inline bool ends_with(const std::basic_string<T>& str, const std::basic_string<T>& suffix) {
+    if (str.length() >= suffix.length()) {
+        return (0 == str.compare(str.length() - suffix.length(), suffix.length(), suffix));
+    }
+    return false;
```
Comment (ilyachur) [inline src/common/util/include/openvino/util/common_util.hpp:107]:
```suggestion
    return str.length() >= suffix.length() && 0 == str.compare(str.length() - suffix.length(), suffix.length(), suffix);
```

File: src/core/tests/frontend/onnx/op_extension.cpp [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+
+class Relu1 : public Relu {
+public:
+    OPENVINO_OP("CustomRelu_1");
+    OPENVINO_FRAMEWORK_MAP(onnx)
+};
+
+class Relu2 : public Relu {
+public:
+    OPENVINO_FRAMEWORK_MAP(onnx, "CustomRelu_2")
```
Comment (ilyachur) [inline src/core/tests/frontend/onnx/op_extension.cpp:27]: Should you add `OPENVINO_OP` also?

Comment (itikhono) [reply src/core/tests/frontend/onnx/op_extension.cpp:27]: not necessary, in this case we will use the base class name for OpenVINO op and "CustomRelu_2" for framework op to handle it in FrontEnd

File: src/bindings/python/src/pyopenvino/frontend/tensorflow/extension.cpp (hunk 14 above)
Comment (itikhono) [reply src/bindings/python/src/pyopenvino/frontend/tensorflow/extension.cpp:12]: most of the files already have the correct copyrights, I updated the few remaining ones.

File: src/bindings/python/src/pyopenvino/frontend/extension.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply src/bindings/python/src/pyopenvino/frontend/extension.cpp:133]: Yes, you are right, we should explicitly cast types here.
I guess py::cast is enough for c++ built-in types and std lib types , and I don't think that proto file will contain not standard types for attributes. 
So, do you think PyAny is a right place to handle these type conversions? Can we cover it in the separate as a bug with high priority?

**Review by ilyachur — APPROVED** [review]

File: src/bindings/python/src/pyopenvino/frontend/extension.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply src/bindings/python/src/pyopenvino/frontend/extension.cpp:133]: Discussed offline, we decided to create a new high priority bug to handle this issue

File: src/bindings/python/src/pyopenvino/frontend/onnx/extension.cpp [hunk 17]
```diff
@@ -41,3 +42,37 @@ void regclass_frontend_onnx_ConversionExtension(py::module m) {
 
     ext.def_property_readonly_static("m_converter", &ConversionExtension::get_converter);
 }
+
+void regclass_frontend_onnx_OpExtension(py::module m) {
+    py::class_<OpExtension<void>, std::shared_ptr<OpExtension<void>>, ConversionExtension> ext(
```
Comment (rkazants) [inline src/bindings/python/src/pyopenvino/frontend/onnx/extension.cpp:47]: OpExtension::Ptr?

File: src/bindings/python/src/pyopenvino/frontend/onnx/extension.cpp [hunk 18]
```diff
@@ -41,3 +42,37 @@ void regclass_frontend_onnx_ConversionExtension(py::module m) {
...
+            m,
+            "OpExtensionONNX",
```
Comment (rkazants) [inline src/bindings/python/src/pyopenvino/frontend/onnx/extension.cpp:49]: Why do we have implemented only `OpExtensionONNX` only? What is about TF and PDPD? Or why don't we have common implementation for all frontends?

File: src/bindings/python/tests/test_frontend/test_frontend_onnx.py [hunk 19]
```diff
@@ -423,3 +423,57 @@ def custom_converter(node: NodeContext):
...
+                                 {"ov_attribute_str": "string",
+                                  "ov_attribute_int": 4,
+                                  "ov_attribute_bool": True,
+                                  "ov_attribute_float": 4.,
+                                  "ov_attribute_vec_string": ["str1", "str2", "str3"],
+                                  "ov_attribute_vec_int": [1, 2, 3, 4, 5, 6, 7],
+                                  "ov_attribute_vec_bool": [True, False, True],
+                                  "ov_attribute_vec_float": [1., 2., 3., 4., 5., 6., 7.]}))
+
+    model = ie.read_model(onnx_model_filename)
```
Comment (rkazants) [inline src/bindings/python/tests/test_frontend/test_frontend_onnx.py:451]: does tested model have those fw operations for which you registered extensions?

Comment (itikhono) [reply src/bindings/python/tests/test_frontend/test_frontend_onnx.py:451]: yes, exactly, it's the same bug as https://github.com/openvinotoolkit/openvino/pull/9917#discussion_r799460031
I will create a new ticket for it

File: src/bindings/python/src/pyopenvino/frontend/onnx/extension.cpp (hunk 18 above)
Comment (itikhono) [reply src/bindings/python/src/pyopenvino/frontend/onnx/extension.cpp:49]: not only onnx, we also have OpExtensionTensorflow and frontend::OpExtension (common)
PaddlePaddle has a specific, we need to map named inputs and named outputs to ov inputs/outputs, so we can't use it without additional information from user, e.g. edges

File: src/bindings/python/src/pyopenvino/frontend/onnx/extension.cpp (hunk 17 above)
Comment (itikhono) [reply src/bindings/python/src/pyopenvino/frontend/onnx/extension.cpp:47]: we don't have default Ptr for onnx::OpExtension, we can use onnx::OpExtension with different template types

File: src/frontends/common/include/openvino/frontend/extension/op.hpp [hunk 20]
```diff
@@ -0,0 +1,248 @@
...
+#include "openvino/core/extension.hpp"
+#include "openvino/frontend/extension/conversion.hpp"
+#include "openvino/frontend/node_context.hpp"
+#include "openvino/frontend/visibility.hpp"
+#include "openvino/opsets/opset.hpp"
+
+namespace ov {
+namespace frontend {
+
+inline const ov::OpSet& get_opset_by_name(const std::string& opset_name) {
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/frontends/common/include/openvino/frontend/extension/op.hpp:23]: done

File: src/core/include/openvino/core/attribute_adapter.hpp [hunk 21]
```diff
@@ -94,6 +107,18 @@ class IndirectScalarValueAccessor : public ValueAccessor<VAT> {
         m_buffer_valid = false;
     }
 
+    virtual void set_as_any(const ov::Any& x) override {
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/core/include/openvino/core/attribute_adapter.hpp]: done

File: src/bindings/python/src/pyopenvino/frontend/onnx/extension.cpp (hunk 18 above)
Comment (itikhono) [reply src/bindings/python/src/pyopenvino/frontend/onnx/extension.cpp:49]: see pyopenvino/frontend/extension.cpp/hpp
pyopenvino/frontend/tensorflow/extension.cpp/hpp

File: src/core/tests/frontend/tensorflow/op_extension.cpp [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+#include "tf_utils.hpp"
+
+using namespace ov::frontend;
+
+using TFOpExtensionTest = FrontEndOpExtensionTest;
+
+class Relu1 : public Relu {
+public:
+    OPENVINO_OP("CustomRelu_1");
+    OPENVINO_FRAMEWORK_MAP(tensorflow)
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/core/tests/frontend/tensorflow/op_extension.cpp:20]: what do you mean? We already use shared tests. Test implementations (TEST_P) and Base class of Relu op are located in shared tests.
We do not link tensorflow_frontend to shared test component, tensorflow part should be tested separately and should be modular

File: src/bindings/python/tests/test_frontend/test_frontend_onnx.py (hunk 19 above)
Comment (itikhono) [reply src/bindings/python/tests/test_frontend/test_frontend_onnx.py:451]: ticket: 78135

File: src/bindings/python/src/pyopenvino/frontend/extension.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply src/bindings/python/src/pyopenvino/frontend/extension.cpp:133]: ticket: 78135

**Review by jiwaszki — APPROVED** [review]
Other than `ov::Any` conversion fix, it is LGTM.

**Review by slyalin — APPROVED** [review]

**Review by rkazants — APPROVED** [review]
